File: fridaLibPatcher.js
/// <reference path="frida-gum.d.ts" />

const Libcam = {

    init() {
 
        //Find GCam lib. Make sure you have libpatcher off!
            let gclib = Process.findModuleByName("libgcastartup.so");

        //Calculate lib size
            Libcam.begin = gclib.base;
            Libcam.size = gclib.size;
            Libcam.end = Libcam.begin.add(Libcam.size);

        //Echo values to console for debugging
            // console.log('Lib stats:')
            // console.log('Size (bytes):' + gclib.size);
            // console.log('Base address:' + gclib.base);
            // console.log('End address:' + Libcam.end);

    },

    offset(addr) {

        //Calculate new hex address for libpatcher value 
            let newoff = Libcam.begin.add(addr) 

        //Echo new offset for debugging
            console.log('New offset of libpatcher value is:' + newoff)

        //Return new value to routine
            return newoff;
    },

    size() {
        //Return gcam lib size
            return gclib.size
    }

};

const Armceptor = {

	writemem: function(addr,newhex) {
		Memory.patchCode(addr, 4, code => {
            const writer = new Arm64Writer(code, { pc: addr });

            writer.putInstruction(newhex); //Replace with newhex when confirmed working
            writer.flush();
        })
	},

	readmem: function(addr,size) {
        let results = hexdump(addr, {
            offset: 0,
            length: size,
            header: false,
            ansi: false
            });
        console.log(results);
    }

};

function swap32(val) {
    return ((val & 0xFF) << 24)
           | ((val & 0xFF00) << 8)
           | ((val >> 8) & 0xFF00)
           | ((val >> 24) & 0xFF);
};

const LibPatcher = {
    patch_hex(libaddr, newhex) {

        libaddr = Number('0x' + libaddr)
        newhex = swap32(Number('0x' + newhex))

        Armceptor.writemem(Libcam.offset(libaddr),newhex);
        Armceptor.readmem(Libcam.offset(libaddr),4);
    }
}



function patchlibaddrs(libaddr, newhex) {

    Libcam.init();
    LibPatcher.patch_hex(libaddr,newhex);
}

function getPhotoName() {
    Interceptor.attach(Module.findExportByName("libgcastartup.so", ".fileno"), {
        onEnter: function(args) {
          this.flag = false;
          var filename = Memory.readCString(ptr(args[0]));
          console.log('filename =', filename)
          if (filename.endsWith(".jpg")) {
            this.flag = true;
            var backtrace = Thread.backtrace(this.context, Backtracer.ACCURATE).map(DebugSymbol.fromAddress).join("\n\t");
            console.log("file name [ " + Memory.readCString(ptr(args[0])) + " ]\nBacktrace:" + backtrace);
          }
        },
        onLeave: function(retval) {
          if (this.flag) // passed from onEnter
            console.warn("\nretval: " + retval);
        }
      });
}

rpc.exports = {
    libpatcher: patchlibaddrs,
    getphotoname: getPhotoName
  // the name of the export (callsecretfunction) cannot have either uppercase letter nor underscores.
};

File: fridaLibPatcher.py
import frida
import time
from ProjectPepega import arm as hex2arm
import sys

#Define a message handler so Python can communicate with the remote Frida instance.
def msg_handler(message, payload):
    print(message)
    print(payload)

# Javascript defaults to Big Endian so the bits in the libpatcher value appear backwards. 
# It's much easier to fix it here than in JavapatchScript.
def swap_endianness(hexstring):
	ba = bytearray.fromhex(hexstring)
	ba.reverse()
	return ba.hex()

# Function to start the remote Frida instance and patch the lib
def load_hook(packageName):
    
	# ADB - initialise interface (to pull image file later)
	# Frida - Connect to device and spawn package
	device = frida.get_usb_device()
	pid = device.spawn([packageName])
	device.resume(pid)

	#Without waiting Java.perform silently fails
	time.sleep(1)
														
	session = device.attach(pid)

	# Create a new JS rampatcher script from string.
	patchScript = session.create_script("""
			function writemem(libOffset,newHex,sizeHex) {
				const libRamAddress = Process.findModuleByName("libgcastartup.so");
				const ramOffset = new NativePointer(libRamAddress.base.add("0x" + libOffset));
				const patchedHex = Number('0x' + newHex);
				const sizeInBytes = Number(sizeHex);

				Memory.patchCode(ramOffset,sizeInBytes,code => {
					const writer = new Arm64Writer(code, ramOffset);
					writer.putInstruction(patchedHex);
					writer.flush();
				});

				Interceptor.attach(
					Module.findExportByName("libc.so", "open"), {
						onEnter: function (args) {
							
							var filePath = Memory.readCString(args[0]);

							if (filePath.includes("/storage/emulated/0/DCIM/Camera/") || 
       							filePath.includes("/storage/emulated/0/Pictures/Raw" )){
								send(filePath);
							} 
						},
						onLeave: function (retval) {
						}
					}
				);
			}

			rpc.exports = {
						libpatcher: writemem
			};
	""")

	patchScript.on('message',msg_handler) 									#Calls message handler for JS
	patchScript.load()
 
	return patchScript

def patch_and_snap(patchScript,libParams):
    #Loop through and patch for each tunable item in the API
	for tunable in libParams:
		hexAddress = tunable['address']
		hexOriginal = tunable['hex_original']
		newValue = tunable['range'][1]
		hexSize = tunable['length_in_lib']

		hexNew = hex2arm.generate_hex(hexOriginal, newValue)
		if hexNew == None:
			hexNew = hex(int(newValue))
		else:
			hexNew = swap_endianness(hexNew)

		#Patch values in RAM
		print("Patching ")
		patchScript.exports.libpatcher(hexAddress, hexNew, hexSize)
	